openedx_certificates/models.py

```diff
@@ -1,5 +1,345 @@
-"""
-Database models for openedx_certificates.
-"""
-# from django.db import models
-# from model_utils.models import TimeStampedModel
+"""Database models for openedx_certificates."""
+from __future__ import annotations
+
+import json
+import logging
+import uuid
+from importlib import import_module
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import jsonfield
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from django_celery_beat.models import IntervalSchedule, PeriodicTask
+from model_utils.models import TimeStampedModel
+from opaque_keys.edx.django.models import CourseKeyField
+
+from openedx_certificates.exceptions import AssetNotFoundError, CertificateGenerationError
+
+if TYPE_CHECKING:  # pragma: no cover
+    from django.core.files import File
+    from django.db.models import QuerySet
+
+
+log = logging.getLogger(__name__)
+
+
+class ExternalCertificateType(TimeStampedModel):
+    """
+    Model to store global certificate configurations for each type.
+
+    .. no_pii:
+    """
+
+    name = models.CharField(max_length=255, unique=True, help_text=_('Name of the certificate type.'))
+    retrieval_func = models.CharField(max_length=200, help_text=_('A name of the function to retrieve eligible users.'))
+    generation_func = models.CharField(max_length=200, help_text=_('A name of the function to generate certificates.'))
+    custom_options = jsonfield.JSONField(default=dict, blank=True, help_text=_('Custom options for the functions.'))
+
+    # TODO: Document how to add custom functions to the certificate generation pipeline.
+
+    def __str__(self):
+        """Get a string representation of this model's instance."""
+        return self.name
+
+    def clean(self):
+        """Ensure that the `retrieval_func` and `generation_func` exist."""
+        for func_field in ['retrieval_func', 'generation_func']:
+            func_path = getattr(self, func_field)
+            try:
+                module_path, func_name = func_path.rsplit('.', 1)
+                module = import_module(module_path)
+                getattr(module, func_name)  # Will raise AttributeError if the function does not exist.
```

Comment:
Nit: I think it would be nice if the above three lines to load the actual function (for `retrieval_func` and `generation_func`) were implemented as properties or methods on this class. They are currently duplicated in `get_eligible_user_ids` and `generate_certificate_for_user` for example.